File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/services/TopUpConfirmation.java
package it.giannibombelli.wsc2026.giftcard.application.services;

import it.giannibombelli.wsc2026.giftcard.application.commands.ConfirmTopUp;
import it.giannibombelli.wsc2026.giftcard.application.usecases.TopUpConfirming;
import it.giannibombelli.wsc2026.giftcard.domain.policies.ConfirmTopUpPolicy;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;

import static java.util.Objects.requireNonNull;


public class TopUpConfirmation {
    private final ConfirmTopUpPolicy policy;
    private final TopUpConfirming useCase;

    public TopUpConfirmation(ConfirmTopUpPolicy paymentPolicy, TopUpConfirming paymentConfirmation) {
        this.policy = requireNonNull(paymentPolicy);
        this.useCase = requireNonNull(paymentConfirmation);
    }

    public void handlePaymentResults(PaymentResultEvents event) {
        final ConfirmTopUp cmd = policy.evaluate(event);

        if (cmd != null) {
            useCase.invoke(cmd);
        }
    }
}

File: typescript/src/giftcard/application/services/topUpConfirmation.ts
// Servizio applicativo che reagisce agli esiti dei pagamenti e conferma la ricarica.
// Violazione cross-BC didattica: GiftCard non dovrebbe importare tipi da Payment.

import type { PaymentResultEvent } from '@/payment/domain/events/paymentResultEvents.js';
import { TopUpConfirming } from '../usecases/topUpConfirming.js';
import { ConfirmTopUpPolicy } from '../../domain/policies/confirmTopUpPolicy.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class TopUpConfirmation {
  private readonly _policy: ConfirmTopUpPolicy;
  private readonly _useCase: TopUpConfirming;

  constructor(policy: ConfirmTopUpPolicy, useCase: TopUpConfirming) {
    requireDependency(policy, "policy");
    requireDependency(useCase, "useCase");
    this._policy = policy;
    this._useCase = useCase;
  }

  handlePaymentResults(event: PaymentResultEvent): void {
    const command = this._policy.evaluate(event);
    if (command !== null) {
      this._useCase.invoke(command);
    }
  }
}
